libnetwork/idm/idm.go

```diff
@@ -4,40 +4,36 @@ package idm
 import (
 	"errors"
 	"fmt"
+	"sync"
 
-	"github.com/docker/docker/libnetwork/bitseq"
-	"github.com/docker/docker/libnetwork/datastore"
+	"github.com/docker/docker/libnetwork/bitmap"
 )
 
-// Idm manages the reservation/release of numerical ids from a contiguous set
+// Idm manages the reservation/release of numerical ids from a contiguous set.
+// Idm is safe for concurrent use.
 type Idm struct {
 	start  uint64
 	end    uint64
-	handle *bitseq.Handle
+	handle *bitmap.Bitmap
+	mu     sync.Mutex
 }
 
 // New returns an instance of id manager for a [start,end] set of numerical ids
-func New(ds datastore.DataStore, id string, start, end uint64) (*Idm, error) {
-	if id == "" {
-		return nil, errors.New("Invalid id")
-	}
+func New(start, end uint64) (*Idm, error) {
 	if end <= start {
 		return nil, fmt.Errorf("Invalid set range: [%d, %d]", start, end)
 	}
 
-	h, err := bitseq.NewHandle("idm", ds, id, 1+end-start)
-	if err != nil {
-		return nil, fmt.Errorf("failed to initialize bit sequence handler: %s", err.Error())
-	}
-
-	return &Idm{start: start, end: end, handle: h}, nil
+	return &Idm{start: start, end: end, handle: bitmap.New(1 + end - start)}, nil
 }
 
 // GetID returns the first available id in the set
 func (i *Idm) GetID(serial bool) (uint64, error) {
 	if i.handle == nil {
 		return 0, errors.New("ID set is not initialized")
 	}
+	i.mu.Lock()
+	defer i.mu.Unlock()
```

Comment:
Should the lock be done before accessing `i.handle` ? Or is it only theoretical for `i.handle` to be mutated?

```suggestion
	i.mu.Lock()
	defer i.mu.Unlock()
	if i.handle == nil {
		return 0, errors.New("ID set is not initialized")
	}
```

Reply:
The `i.handle` check guards against panics if the methods are called on a zero-valued idm. The struct field is immutable after construction, and therefore does not need to be guarded by a mutex.
```go
var i idm.Idm // uninitialized
i.GetID(true) // bad
```